api/envoy/config/route/v3/route_components.proto

```diff
@@ -837,6 +837,62 @@ message RouteAction {
   string prefix_rewrite = 5
       [(validate.rules).string = {well_known_regex: HTTP_HEADER_VALUE strict: false}];
 
+  // Indicates that some headers received from upstream cluster need to be rewritten. When upstream
+  // returns a header, the header will be overwritten with the given value before being sent back as
+  // response to the client.
+  //
+  // In the below example there are two response header rewrites mentioned: one for location header
+  // and one for x-random-token header.
+  //
+  // The location header's value will be matched with the given match pattern as a prefix to the header's
+  // value. If a match exists, the matched portion may be swapped with the value given in substitution.
+  // If no match string is given, the value is replaced entirely with the given substitution. The
+  // response status header is also matched before making a substitution. If the response status
+  // returned from the upstream cluster matches with one of the response statuses in the list, the
+  // substitution can proceed after matching the match section. If this list is empty, response status
```

Comment:
I think the response status is really part of the matcher. Also, this only applies to the literal and not when we are doing regex substitute. I think you might want to move this up a level to a `header_rewriter` message.